File: 06-state-management/golang/agent.go
package main

import (
	"context"
	"fmt"
	"os"
	"strconv"
	"strings"
	"sync"

	"byoo/xclaw/providers"
)

const defaultMaxIterations = 10

// Agent manages per-session conversation history and drives the agent loop.
// db may be nil (in-memory only mode); when set, all sessions and traces are
// persisted to SQLite for crash recovery, reconnect, rollback and fork.
type Agent struct {
	mu            sync.Mutex
	sessions      map[string][]providers.Message // in-memory cache, write-through to db
	db            *DB
	providerChain []string
	maxIterations int
}

func newAgent(providerChain []string, maxIterations int, db *DB) *Agent {
	if maxIterations <= 0 {
		maxIterations = defaultMaxIterations
	}
	return &Agent{
		sessions:      make(map[string][]providers.Message),
		db:            db,
		providerChain: providerChain,
		maxIterations: maxIterations,
	}
}

// handle processes one user message for the given session.
// onDelta is called for each streamed token once we confirm the reply is plain text.
//
// Special commands (handled before the LLM loop):
//   /rollback <stepID>          — delete traces after stepID, reset session
//   /fork <stepID> [new title]  — clone session up to stepID into a new session
func (a *Agent) handle(msg ACPMessage, onDelta func(string)) (string, error) {
	// ── Special commands ──────────────────────────────────────────────────────
	if a.db != nil {
		if reply, handled := a.handleCommand(msg); handled {
			onDelta(reply)
			return reply, nil
		}
	}

	// ── Load or initialise session ────────────────────────────────────────────
	a.mu.Lock()
	if _, ok := a.sessions[msg.SessionID]; !ok {
		// Cache miss: try to restore from DB (server-restart / reconnect case).
		if a.db != nil {
			if status := a.db.getStatus(msg.SessionID); status != "" {
				loaded, err := a.db.loadMessages(msg.SessionID)
				if err == nil {
					a.sessions[msg.SessionID] = loaded
					fmt.Fprintf(os.Stderr, "[agent] [%s] restored %d messages from DB (status: %s)\n",
						msg.SessionID, len(loaded), status)
				}
			}
		}
		// Still empty → brand-new session.
		if _, ok := a.sessions[msg.SessionID]; !ok {
			a.sessions[msg.SessionID] = []providers.Message{
				{Role: "system", Content: buildSystemPrompt()},
			}
		}
	}
	messages := make([]providers.Message, len(a.sessions[msg.SessionID]))
	copy(messages, a.sessions[msg.SessionID])
	a.mu.Unlock()

	// ── Persist session + set Running ─────────────────────────────────────────
	// State is committed to DB BEFORE any action is taken.
	// On crash, the DB shows Running so reconnect can resume instead of replaying.
	if a.db != nil {
		a.db.ensureSession(msg.SessionID, "Session "+msg.SessionID)
		a.db.setStatus(msg.SessionID, StatusRunning)
	}

	// ── Record user_input trace ───────────────────────────────────────────────
	var lastStepID string
	if a.db != nil {
		sid, _ := a.db.beginTrace(msg.SessionID, "", "user_input", "user", msg.Content)
		a.db.completeTrace(sid, msg.Content, 0, 0)
		lastStepID = sid
	}

	messages = append(messages, providers.Message{Role: "user", Content: msg.Content})

	// ── Agent loop ────────────────────────────────────────────────────────────
	for iterations := 0; iterations < a.maxIterations; iterations++ {
		// Begin llm_call trace BEFORE calling the LLM (state-before-action).
		var llmStepID string
		if a.db != nil {
			llmStepID, _ = a.db.beginTrace(msg.SessionID, lastStepID, "llm_call", "llm", nil)
		}

		var buffer []string
		reply, err := providers.StreamWithFallback(
			context.Background(), messages, a.providerChain, assemble,
			func(token string) { buffer = append(buffer, token) },
		)
		if err != nil {
			if a.db != nil && llmStepID != "" {
				a.db.failTrace(llmStepID, err.Error())
				a.db.setStatus(msg.SessionID, StatusFailed)
			}
			return "", err
		}

		// Complete llm_call trace with the raw reply.
		if a.db != nil && llmStepID != "" {
			a.db.completeTrace(llmStepID, reply, 0, 0)
		}
		messages = append(messages, providers.Message{Role: "assistant", Content: reply})
		lastStepID = llmStepID

		toolCall := extractJSON(reply)
		action, isToolCall := "", false
		if toolCall != nil {
			if act, ok := toolCall["action"].(string); ok && act != "" {
				action = act
				isToolCall = true
			}
		}

		if !isToolCall {
			// Plain text reply — flush buffered tokens to the channel adapter.
			for _, t := range buffer {
				onDelta(t)
			}
			if a.db != nil {
				a.db.setStatus(msg.SessionID, StatusSuccess)
			}
			a.mu.Lock()
			a.sessions[msg.SessionID] = messages
			a.mu.Unlock()
			return reply, nil
		}

		tool, found := toolRegistry[action]
		if !found {
			errMsg := fmt.Sprintf(`error: unknown tool "%s". Available: %s`, action, availableTools())
			messages = append(messages, providers.Message{Role: "user", Content: errMsg})
			continue
		}

		params := make(map[string]string)
		for k, v := range toolCall {
			if k != "action" {
				if s, ok := v.(string); ok {
					params[k] = s
				}
			}
		}

		// Begin tool_call trace BEFORE executing the tool (state-before-action).
		// If the process crashes here, DB shows a "running" tool_call with no output —
		// reconnect can detect this and retry or skip it.
		var toolStepID string
		if a.db != nil {
			toolStepID, _ = a.db.beginTrace(msg.SessionID, llmStepID, "tool_call", action, params)
		}

		fmt.Fprintf(os.Stderr, "[agent] [%s] [%s] uses [%s]: %v\n", msg.SessionID, toolStepID, action, params)
		output, execErr := tool.Execute(msg.SessionID, params)
		if execErr != nil {
			fmt.Fprintf(os.Stderr, "[agent] tool error: %v\n", execErr)
			if a.db != nil && toolStepID != "" {
				a.db.failTrace(toolStepID, execErr.Error())
			}
			messages = append(messages, providers.Message{Role: "user", Content: "tool error:\n" + execErr.Error()})
		} else {
			if a.db != nil && toolStepID != "" {
				a.db.completeTrace(toolStepID, output, 0, 0)
			}
			messages = append(messages, providers.Message{Role: "user", Content: "tool output:\n" + output})
		}
		lastStepID = toolStepID
	}

	notice := fmt.Sprintf("[xclaw] reached max iterations (%d), stopping", a.maxIterations)
	if a.db != nil {
		a.db.setStatus(msg.SessionID, StatusFailed)
	}
	a.mu.Lock()
	a.sessions[msg.SessionID] = messages
	a.mu.Unlock()
	return notice, nil
}

// handleCommand processes /steps, /rollback and /fork slash commands.
// Returns (reply, true) when a command was handled; ("", false) otherwise.
func (a *Agent) handleCommand(msg ACPMessage) (string, bool) {
	content := strings.TrimSpace(msg.Content)

	// /steps [n] — list the most recent n steps (default 10)
	if content == "/steps" || strings.HasPrefix(content, "/steps ") {
		n := 10
		if after, ok := strings.CutPrefix(content, "/steps "); ok {
			if v, err := strconv.Atoi(strings.TrimSpace(after)); err == nil && v > 0 {
				n = v
			}
		}
		steps, err := a.db.loadSteps(msg.SessionID, n)
		if err != nil {
			return "steps failed: " + err.Error(), true
		}
		if len(steps) == 0 {
			return "no steps recorded yet in this session.", true
		}
		// Reverse to show oldest first (loadSteps returns newest first).
		for i, j := 0, len(steps)-1; i < j; i, j = i+1, j-1 {
			steps[i], steps[j] = steps[j], steps[i]
		}
		var sb strings.Builder
		fmt.Fprintf(&sb, "步骤列表 (session: %s):\n", msg.SessionID)
		for _, s := range steps {
			fmt.Fprintf(&sb, "  %-6s  %-12s  %-10s  %s\n", s.StepID, s.StepType, s.Name, s.Snippet)
		}
		sb.WriteString("\n用法:\n")
		sb.WriteString("  /rollback <stepID>        回到该步骤之前重新执行\n")
		sb.WriteString("  /fork <stepID> [title]    从该步骤分叉新会话（原会话保留）\n")
		return sb.String(), true
	}

	// /rollback <stepID>
	// Deletes all traces from stepID onwards and resets session to Running.
	// Clears the in-memory cache so the next message reloads from DB.
	if after, ok := strings.CutPrefix(content, "/rollback "); ok {
		stepID := expandStepID(msg.SessionID, strings.TrimSpace(after))
		if err := a.db.Rollback(msg.SessionID, stepID); err != nil {
			return "rollback failed: " + err.Error(), true
		}
		a.mu.Lock()
		delete(a.sessions, msg.SessionID) // force reload from DB on next handle()
		a.mu.Unlock()
		return fmt.Sprintf("[rollback] session reset to before step %s — send your new instruction", stepID), true
	}

	// /fork <stepID> [new title]
	// Clones session up to stepID into a new session ID. Original is untouched.
	if after, ok := strings.CutPrefix(content, "/fork "); ok {
		parts := strings.SplitN(strings.TrimSpace(after), " ", 2)
		stepID := expandStepID(msg.SessionID, parts[0])
		title := "fork from " + stepID
		if len(parts) > 1 {
			title = strings.TrimSpace(parts[1])
		}
		newSessionID := "fork_" + randomHex(6)
		if err := a.db.Fork(msg.SessionID, newSessionID, title, stepID); err != nil {
			return "fork failed: " + err.Error(), true
		}
		return fmt.Sprintf(
			"[fork] new session created: %s\n"+
				"connect with this session ID to continue on the forked branch.\n"+
				"original session %s is unchanged.",
			newSessionID, msg.SessionID,
		), true
	}

	return "", false
}

File: 06-state-management/nodejs/src/agent.ts
import { streamWithFallback } from './providers/registry.ts';
import { toolRegistry, buildToolsPrompt, extractJSON } from './tools.ts';
import type { Message } from './providers/types.ts';
import type { ACPMessage } from './gateway/types.ts';
import { log } from './logger.ts';
import { DB, expandStepId, shortStepId } from './db.ts';

export class Agent {
  private sessions = new Map<string, Message[]>();
  private providerChain: string[];
  private maxIterations: number;
  private db: DB | null;

  constructor(providerChain: string[], maxIterations: number, db: DB | null = null) {
    this.providerChain = providerChain;
    this.maxIterations = maxIterations;
    this.db = db;
  }

  // Process one user message; call onDelta for each streamed token.
  async handle(msg: ACPMessage, onDelta: (token: string) => void): Promise<string> {
    // ── Special commands ──────────────────────────────────────────────────────
    if (this.db) {
      const reply = this.handleCommand(msg);
      if (reply !== null) {
        onDelta(reply);
        return reply;
      }
    }

    // ── Load or initialise session ────────────────────────────────────────────
    if (!this.sessions.has(msg.sessionId)) {
      // Cache miss: try to restore from DB (server-restart / reconnect case).
      if (this.db) {
        const status = this.db.getStatus(msg.sessionId);
        if (status !== null) {
          const loaded = this.db.loadMessages(msg.sessionId, buildSystemPrompt());
          this.sessions.set(msg.sessionId, loaded);
          log(`[agent] [${msg.sessionId}] restored ${loaded.length} messages from DB (status: ${status})`);
        }
      }
      if (!this.sessions.has(msg.sessionId)) {
        this.sessions.set(msg.sessionId, [{ role: 'system', content: buildSystemPrompt() }]);
      }
    }
    const messages = [...this.sessions.get(msg.sessionId)!];

    // ── Persist session + set Running ─────────────────────────────────────────
    if (this.db) {
      this.db.ensureSession(msg.sessionId, 'Session ' + msg.sessionId);
      this.db.setStatus(msg.sessionId, 'Running');
    }

    // ── Record user_input trace ───────────────────────────────────────────────
    let lastStepId = '';
    if (this.db) {
      const sid = this.db.beginTrace(msg.sessionId, '', 'user_input', 'user', msg.content);
      this.db.completeTrace(sid, msg.content);
      lastStepId = sid;
    }

    messages.push({ role: 'user', content: msg.content });

    // ── Agent loop ────────────────────────────────────────────────────────────
    let iterations = 0;
    while (true) {
      if (++iterations > this.maxIterations) {
        const notice = `[xclaw] reached max iterations (${this.maxIterations}), stopping`;
        if (this.db) this.db.setStatus(msg.sessionId, 'Failed');
        log(notice);
        this.sessions.set(msg.sessionId, messages);
        return notice;
      }

      // Begin llm_call trace BEFORE calling the LLM (state-before-action).
      let llmStepId = '';
      if (this.db) {
        llmStepId = this.db.beginTrace(msg.sessionId, lastStepId, 'llm_call', 'llm', null);
      }

      const buffer: string[] = [];
      let reply: string;
      try {
        reply = await streamWithFallback(messages, this.providerChain, (token) => {
          buffer.push(token);
        });
      } catch (err: any) {
        if (this.db && llmStepId) {
          this.db.failTrace(llmStepId, err.message);
          this.db.setStatus(msg.sessionId, 'Failed');
        }
        throw err;
      }

      if (this.db && llmStepId) {
        this.db.completeTrace(llmStepId, reply);
      }
      messages.push({ role: 'assistant', content: reply });
      lastStepId = llmStepId;

      const toolCall = extractJSON(reply);
      if (toolCall && typeof toolCall['action'] === 'string') {
        const action = toolCall['action'];
        const tool = toolRegistry.get(action);
        if (tool) {
          const params: Record<string, string> = {};
          for (const [k, v] of Object.entries(toolCall as Record<string, string>)) {
            if (k !== 'action' && typeof v === 'string') params[k] = v;
          }

          // Begin tool_call trace BEFORE executing (state-before-action).
          let toolStepId = '';
          if (this.db) {
            toolStepId = this.db.beginTrace(msg.sessionId, llmStepId, 'tool_call', action, params);
          }

          log(`[agent] [${msg.sessionId}] [${toolStepId || '-'}] uses [${action}]: ${JSON.stringify(params)}`);
          try {
            const output = await tool.execute(msg.sessionId, params);
            log(output);
            if (this.db && toolStepId) this.db.completeTrace(toolStepId, output);
            messages.push({ role: 'user', content: `tool output:\n${output}` });
          } catch (err: any) {
            const errMsg = err.stderr ?? err.message;
            console.error(`[agent] tool error: ${errMsg}`);
            if (this.db && toolStepId) this.db.failTrace(toolStepId, errMsg);
            messages.push({ role: 'user', content: `tool error:\n${errMsg}` });
          }
          lastStepId = toolStepId || llmStepId;
        } else {
          const available = [...toolRegistry.keys()].join(', ');
          messages.push({ role: 'user', content: `error: unknown tool "${action}". Available: ${available}` });
        }
      } else {
        // Plain text reply — flush buffered tokens to the channel adapter.
        for (const token of buffer) onDelta(token);
        if (this.db) this.db.setStatus(msg.sessionId, 'Success');
        this.sessions.set(msg.sessionId, messages);
        return reply;
      }
    }
  }

  // handleCommand processes /steps, /rollback, /fork slash commands.
  // Returns the reply string if a command was handled; null otherwise.
  private handleCommand(msg: ACPMessage): string | null {
    const content = msg.content.trim();

    // /steps [n] — list the most recent n steps (default 10)
    if (content === '/steps' || content.startsWith('/steps ')) {
      let n = 10;
      const after = content.slice('/steps'.length).trim();
      if (after) {
        const parsed = parseInt(after, 10);
        if (!isNaN(parsed) && parsed > 0) n = parsed;
      }
      const steps = this.db!.loadSteps(msg.sessionId, n);
      if (steps.length === 0) return 'no steps recorded yet in this session.';
      // Reverse to show oldest first (loadSteps returns newest first).
      steps.reverse();
      let out = `步骤列表 (session: ${msg.sessionId}):\n`;
      for (const s of steps) {
        out += `  ${s.stepId.padEnd(6)}  ${s.stepType.padEnd(12)}  ${s.name.padEnd(10)}  ${s.snippet}\n`;
      }
      out += '\n用法:\n';
      out += '  /rollback <stepID>        回到该步骤之前重新执行\n';
      out += '  /fork <stepID> [title]    从该步骤分叉新会话（原会话保留）\n';
      return out;
    }

    // /rollback <stepID>
    if (content.startsWith('/rollback ')) {
      const short = content.slice('/rollback '.length).trim();
      const stepId = expandStepId(msg.sessionId, short);
      try {
        this.db!.rollback(msg.sessionId, stepId);
        this.sessions.delete(msg.sessionId); // force reload from DB on next handle()
        return `[rollback] session reset to before step ${short} — send your new instruction`;
      } catch (err: any) {
        return `rollback failed: ${err.message}`;
      }
    }

    // /fork <stepID> [new title]
    if (content.startsWith('/fork ')) {
      const rest = content.slice('/fork '.length).trim();
      const spaceIdx = rest.indexOf(' ');
      const short = spaceIdx >= 0 ? rest.slice(0, spaceIdx) : rest;
      const title = spaceIdx >= 0 ? rest.slice(spaceIdx + 1).trim() : `fork from ${short}`;
      const stepId = expandStepId(msg.sessionId, short);
      const newSessionId = 'fork_' + crypto.randomUUID().slice(0, 6);
      try {
        this.db!.fork(msg.sessionId, newSessionId, title, stepId);
        return (
          `[fork] new session created: ${newSessionId}\n` +
          `connect with this session ID to continue on the forked branch.\n` +
          `original session ${msg.sessionId} is unchanged.`
        );
      } catch (err: any) {
        return `fork failed: ${err.message}`;
      }
    }

    return null;
  }
}

export function buildSystemPrompt(): string {
  return `You are an AI assistant named xclaw.

To use a tool, output a JSON object (bare or in a markdown code block):
{"action": "<tool_name>", "<param1>": "<value1>", ...}

To answer directly, output plain text — do NOT use JSON.

Available tools:
${buildToolsPrompt()}`;
}
